python/lsst/ts/adamSensors/model.py

```diff
@@ -21,35 +24,38 @@ class AdamModel:
             the pymodbus object representing the ADAM 6024
     """
 
-    def __init__(self, log=None, simulation_mode=False):
+    def __init__(self, ip, port, log=None, simulation_mode=False):
+        self.loop = asyncio.new_event_loop()
+        self.clientip = ip
+        self.clientport = port
+        self.client = None
+
+        self.t = Thread(target=self._start_loop, args=[self.loop])
```

Comment:
I see what you mean about ugly. Oy. It's probably worth some time to clean this up but I'd want to think about the best way to do it.

There ought to be some way to make the ModBus client use your existing asyncio event loop but I know that you tried that.

A fallback is to try `loop.run_in_executor` in order to avoid threading. That's the recommendation I found in the asyncio docs for synchronous code and I think it would work here. For a long-running task, such as this, you can use run it as a background task, e.g.:
```
loop = asyncio.get_running_loop()
self.my_task = asyncio.create_task(loop.run_in_executor(None, my_callable, arg1, arg2. ,,,))
```
You can cancel the task to kill the background job.

(For short-term tasks that you want to wait for, use `await loop.run_in_executor`.)

Note that run_in_executor _only_ accepts positional arguments. If you have enough args that you want to use named args then you can work around this limitation by using functools.partial to wrap the function and call run_in_executor on the wrapped result.

Reply:
That said, if you are going to use threads or run_in_executor it may be simpler to just run the synchronous modbus code. The example here https://pymodbus.readthedocs.io/en/latest/source/example/async_asyncio_client.html (which I suspect is what you are using) is really off-putting as far as an argument for using the async code in pymodbus.

async-modbus looks promising, but it's in pre-alpha state.